lib/AST/ShouldPrintForParseableInterface.cpp

```diff
@@ -0,0 +1,155 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "ShouldPrintForParseableInterface.h"
+#include "swift/AST/ASTWalker.h"
+#include "swift/AST/TypeWalker.h"
+#include "swift/AST/Decl.h"
+#include "swift/AST/Module.h"
+#include "swift/AST/Type.h"
+
+using namespace swift;
+
+bool swift::isPublicOrUsableFromInline(const ValueDecl *VD) {
+  AccessScope scope =
+  VD->getFormalAccessScope(/*useDC*/nullptr,
+                           /*treatUsableFromInlineAsPublic*/true);
+  return scope.isPublic();
+}
+
+bool swift::contributesToParentTypeStorage(const AbstractStorageDecl *ASD) {
+  auto *DC = ASD->getDeclContext()->getAsDecl();
+  if (!DC) return false;
+  auto *ND = dyn_cast<NominalTypeDecl>(DC);
+  if (!ND) return false;
+  return !ND->isResilient() && ASD->hasStorage() && !ASD->isStatic();
+}
+
+/// Finds type decls that need to be printed to ensure the layout of public
+/// types are correct for non-resilient types.
+///
+/// For example:
+///
+/// ```
+/// private struct B {}
+/// public struct A {
+///   internal let b: B
+/// }
+/// ```
+/// Since `A`'s size depends on `B`, we need to print `B` even though it's
+/// private. `B`'s children will also be checked for types that contribute to
+/// its layout, and so on.
+class FindReferencedNonPublicTypeDecls: public ASTWalker {
+  using TypeDecls = SmallPtrSetImpl<TypeDecl *>;
+
+  /// Finds all nominal types or typealiases referenced in the type being
+  /// walked.
+  class FindTypeDecls: public TypeWalker {
+    TypeDecls &foundDecls;
+  public:
+    FindTypeDecls(TypeDecls &foundDecls): foundDecls(foundDecls) {}
+
+    void addIfNonPublic(TypeDecl *typeDecl) {
+      if (isPublicOrUsableFromInline(typeDecl))
+        return;
+      foundDecls.insert(typeDecl);
+    }
+
+    /// Finds all nominal types or typealiases referenced by the
+    Action walkToTypePre(Type type) override {
+      if (auto nominal =
+          dyn_cast<NominalOrBoundGenericNominalType>(type.getPointer()))
+        addIfNonPublic(nominal->getDecl());
+      if (auto alias = dyn_cast<NameAliasType>(type.getPointer())) {
+        // Add the typealias to the found decls
+        auto decl = alias->getDecl();
+        addIfNonPublic(decl);
+
+        // Also walk into the RHS of the typealias to find extra types we need.
+        decl->getUnderlyingTypeLoc().getType().walk(*this);
+      }
+      return Action::Continue;
+    }
+  };
+  FindTypeDecls findTypeDecls;
+public:
+  FindReferencedNonPublicTypeDecls(TypeDecls &foundDecls):
+    findTypeDecls(foundDecls) {}
+  bool walkToDeclPre(Decl *decl) {
```

Comment:
This should

- not walk into decls that aren't going to be printed
- not walk into function bodies

which makes me think you really want a worklist, not a single-pass walk. PrintAsObjC might be a good reference, since it has to do something similar for forward declarations.